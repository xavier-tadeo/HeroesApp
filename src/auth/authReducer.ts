import { types } from "../types/types";

// const state = {
//   name: "arlet",
//   logged: true
// }

type TAction = {
  type: string;
  payload: {
    name: string;
    email: string;
  };
};

export const authReducer = (state = {}, action: TAction) => {
  switch (action.type) {
    case types.login:
      return {
        ...action.payload,
        logged: true,
      };

    case types.logout:
      return {
        logged: false,
      };

    default:
      return state;
  }
};
