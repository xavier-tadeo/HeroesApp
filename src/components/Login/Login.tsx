import React, { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { types } from "../../types/types";

export const Login = () => {
  const navigate = useNavigate();

  const { dispatch } = useContext(AuthContext);

  const handleLogin = () => {
    const action = {
      type: types.login,
      payload: {
        name: "Tadeoo",
        email: "tadeo@tadeo.es",
      },
    };

    dispatch(action);

    const lastPath = localStorage.getItem("lastPath") || "/";
    // const lastSearch = localStorage.getItem("lastSearch" || "");

    navigate(lastPath, {
      replace: true,
    });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
