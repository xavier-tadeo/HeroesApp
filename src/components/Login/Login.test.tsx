import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ReactTestRenderer from "react-test-renderer";
import { AuthContext } from "../../auth/authContext";
import { types } from "../../types/types";
import { Login } from "./Login";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Given a Login component", () => {
  const contextValue = {
    user: {
      name: "",
      logged: false,
    },
    dispatch: jest.fn(),
  };

  describe("When it renderer", () => {
    const view = ReactTestRenderer.create(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={["/login"]}>
          <Login />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    test("Then should render correcty form", () => {
      expect(view.toJSON()).toMatchSnapshot();
    });

    test("Then should do teh dispatch and navigate", () => {
      const action = {
        type: types.login,
        payload: {
          name: "Tadeoo",
          email: "tadeo@tadeo.es",
        },
      };
      localStorage.setItem("lastPath", "/dc");

      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/login"]}>
            <Login />
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const button = screen.getByRole("button");
      fireEvent.click(button);

      expect(contextValue.dispatch).toHaveBeenCalledWith(action);
      expect(mockNavigate).toHaveBeenCalledWith("/dc", { replace: true });
    });
  });
});
