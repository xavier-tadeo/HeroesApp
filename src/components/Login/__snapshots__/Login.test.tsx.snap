// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Login component When it renderer Then should render correcty form 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
